### DISCUSSION TITLE
The repository appears mismanaged

### DISCUSSION THREAD

**Discussion by KyleRudy** [evennia/evennia#3290] [Evennia Engine Development]
Hello! I'm entirely new to evennia, drawn here because some of my colleagues want to port a game to the engine. One thing we all noticed right away was the _massive_ size of the repository, which appears to be mismanaged.

A fresh clone of the repo is 902M, with 885M being .git. This is a symptom of accidentally committing binaries or other build output to the repo when the repo should contain only source.

A quick look with
```
  git rev-list --objects --all |
  git cat-file --batch-check='%(objecttype) %(objectname) %(objectsize) %(rest)' |
  sed -n 's/^blob //p' |
  sort --numeric-sort --key=2 |
  cut -c 1-12,41- |
  $(command -v gnumfmt || echo numfmt) --field=2 --to=iec-i --suffix=B --padding=7 --round=nearest
```

confirms this. We see output dominated by the docs directory. environment.pickle is 9.6 MB on every revision where it changes, and there are plenty of other files in the docs directory that are obviously build output.

I notice that docs/build is in the .gitignore, so this was clearly a known issue at a previous point. However, this .gitignore is missing docs/2.x/* and similar directory structures.

I would suggest that we extend .gitignore to feature all of docs which contains build output, as well as *.pickle. Then, the repo requires a significant maintenance effort to remove the revision history for these binary files from .git.

**Comment by volundmush** [comment]
I've never done git surgery on that kind of scale before, but I'm interested in looking into this. Got any tips on just what I should be looking for or what sort of approach would solve that?

**Reply by KyleRudy** [reply]
If I knew the correct commands to run, I'd share them. Sadly, I've only done git surgery like this twice before in my career, and the last incident was approximately eight years ago. When I have the opportunity to research this further, I will, and if I come up with something to recommend, I won't hesitate to do so.

My personal plan is to start with a fork, then try to clean it up. It's possible such a change will pushable upstream; if not, I'll give you full repro steps.

**Reply by SuperSayu** [reply]
Recently I had to purge files from a much smaller repo.  The tool I found was [git-filter-repo](https://github.com/newren/git-filter-repo), which will retroactively remove files from all git commits:
`git-filter-repo --invert-paths --path <file-to-remove>`
I don't know that I can offer much more than that I tried it and it worked for me

**Reply by KyleRudy** [reply]
This worked on a test repo for me, but it's worth noting that you'll have to trust the tool's output and `git push -f` to overwrite the existing history on each affected branch.

**Comment by Griatch** [comment]
You'll find these artifacts in the repo's `gh-pages` branch. The way github pages works, is that it will pull data directly from this branch to build the documentation / mainsite website. So some 'built' content is necessary and is not a bug per se. 

I'll experiment a bit to see which of these files can be omitted from the final repo though, maybe some are really build artifacts that are not necessary. If not, maybe it may be worthwhile to break the deployment out into its own repo that people are not regularly downloading.

**Comment by Griatch** [comment]
Turns out that the `.doctrees/` subfolder was some sort of extra cache that was not needed to have. So removed that and cleaned up the history, bringing the repo size down from 900M to 220M. :taco:  The main other size-wasters are generated indices and search indices; I suppose one could clean out the history of these too, but they at least don't grow as often.

**Comment by KyleRudy** [comment]
You may wish to add a line like one of these to .gitignore:
```
**/.doctrees
**/*.pickle
```
to handle `docs/2.x/.doctrees/environment.pickle` and anything else that might show up there.

If the `docs/<version_number>` directories are no longer needed, or entirely generated on-demand, you might seriously consider popping them into `.gitignore`. If necessary, add generation instructions to `README.md`.

```
krudy@abc:~/source/evennia$ du -ah | tail -n1
263M    .
krudy@abc:~/source/evennia$ git-filter-repo --invert-paths --path-regex ".*\.pickle" &> /dev/null
krudy@abc:~/source/evennia$ du -ah | tail -n1
260M    .
krudy@abc:~/source/evennia$ git-filter-repo --invert-paths --path-regex "docs/[0-9\.]+" &> /dev/null
krudy@abc:~/source/evennia$ du -ah | tail -n1
79M     .
krudy@abc:~/source/evennia$ git-filter-repo --invert-paths --path "docs/versions" &> /dev/null
krudy@abc:~/source/evennia$ du -ah | tail -n1
75M     .
```

You can probably remove `devblog` from history as well, which reduces it from 75M to 73M.

```
krudy@abc:~/source/evennia$ git-filter-repo --invert-paths --path "devblog" &> /dev/null
krudy@abc:~/source/evennia$ du -ah | tail -n1
73M     .
```

Of that 73M, 15M is the remainder of `docs`, 26M is `evennia`, 14M is `src`, 12M is everything else.

**Reply by KyleRudy** [reply]
Posted this a little prematurely, editing it as I refine it.

**Reply by KyleRudy** [reply]
Alright, I think 73M is small enough for our fork. I'm done editing my disasterpost. Please do with this information what you will, and if by chance those old docs folders are more important than they look, please let me know.
